File: go/internal/xlsxwrite/write.go
// Package xlsxwrite exports a GridMD workbook to an .xlsx package. The
// worksheet core (cell values, formulas + cached values, and merges) is emitted
// natively so the file opens as a real spreadsheet; the complete GridMD source
// is additionally carried, base64-encoded, in customXml/gridmdCarry.xml so that
// every feature the dump measures round-trips losslessly (SPEC §11 — nothing is
// silently dropped). See the README for the deliberate carry design.
package xlsxwrite

import (
	"archive/zip"
	"bytes"
	"encoding/base64"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"

	"github.com/lprhodes/grid-md/go/internal/model"
	"github.com/lprhodes/grid-md/go/internal/numfmt"
	"github.com/lprhodes/grid-md/go/internal/refs"
	"github.com/lprhodes/grid-md/go/internal/scalar"
)

// CarryPart is the fixed package path of the GridMD source carry part.
const CarryPart = "customXml/gridmdCarry.xml"

type part struct {
	name string
	data string
}

// Write renders wb (materialized from source) to an .xlsx byte package and a
// human-readable fidelity report.
func Write(wb *model.Workbook, source []byte) ([]byte, []string, error) {
	parts, report := buildParts(wb, source)
	var buf bytes.Buffer
	// A bytes.Buffer never fails, so err is nil in practice; it is returned
	// directly (no separate branch) to keep this path fully exercised.
	err := writeZip(&buf, parts)
	return buf.Bytes(), report, err
}

// buildParts computes every package part and the fidelity report (pure — no
// I/O), so the byte-emission (writeZip) stays a thin, separately-tested seam.
func buildParts(wb *model.Workbook, source []byte) ([]part, []string) {
	dateSystem := 1900
	if ds, ok := wb.FM["date-system"].(float64); ok && ds == 1904 {
		dateSystem = 1904
	}
	parts := []part{
		{"[Content_Types].xml", contentTypes(len(wb.Sheets))},
		{"_rels/.rels", rootRels()},
		{"xl/workbook.xml", workbookXML(wb)},
		{"xl/_rels/workbook.xml.rels", workbookRels(len(wb.Sheets))},
		{"xl/styles.xml", stylesXML()},
	}
	report := []string{}
	nativeCells := 0
	for i, s := range wb.Sheets {
		xml, cells := worksheetXML(s, dateSystem)
		nativeCells += cells
		parts = append(parts, part{fmt.Sprintf("xl/worksheets/sheet%d.xml", i+1), xml})
		report = append(report, fmt.Sprintf("sheet %q: %d cells, %d merges emitted natively", s.Name, cells, len(s.Merges)))
		if carried := carriedSummary(s); carried != "" {
			report = append(report, "  carried via source part: "+carried)
		}
	}
	parts = append(parts, part{CarryPart, carryXML(base64.StdEncoding.EncodeToString(source))})
	report = append(report,
		fmt.Sprintf("%d cells written natively across %d sheet(s)", nativeCells, len(wb.Sheets)),
		"full GridMD source carried in "+CarryPart+" (lossless round-trip; nothing dropped)")
	return parts, report
}

func writeZip(w io.Writer, parts []part) error {
	zw := zip.NewWriter(w)
	for _, p := range parts {
		// STORE (no compression): keeps parts inspectable and lets every
		// byte reach w directly. archive/zip readers accept STORE and DEFLATE.
		fw, err := zw.CreateHeader(&zip.FileHeader{Name: p.name, Method: zip.Store})
		if err != nil {
			return err
		}
		if _, err := fw.Write([]byte(p.data)); err != nil {
			return err
		}
	}
	return zw.Close()
}

func carriedSummary(s *model.Sheet) string {
	var parts []string
	kv := func(label string, n int) {
		if n > 0 {
			parts = append(parts, fmt.Sprintf("%s=%d", label, n))
		}
	}
	kv("cf", s.CFRules)
	kv("validations", s.Validations)
	kv("notes", s.Notes)
	kv("threads", s.Threads)
	kv("scenarios", s.Scenarios)
	kv("sparklines", s.Sparklines)
	kv("charts", s.Charts)
	kv("pivots", s.Pivots)
	kv("slicers", s.Slicers)
	kv("images", s.Images)
	kv("shapes", s.Shapes)
	kv("hyperlinks", s.Hyperlinks)
	kv("tables", len(s.Tables))
	return strings.Join(parts, " ")
}

func contentTypes(nSheets int) string {
	var b strings.Builder
	b.WriteString(xmlDecl)
	b.WriteString(`<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">`)
	b.WriteString(`<Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>`)
	b.WriteString(`<Default Extension="xml" ContentType="application/xml"/>`)
	b.WriteString(`<Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/>`)
	b.WriteString(`<Override PartName="/xl/styles.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.styles+xml"/>`)
	for i := 0; i < nSheets; i++ {
		fmt.Fprintf(&b, `<Override PartName="/xl/worksheets/sheet%d.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/>`, i+1)
	}
	b.WriteString(`</Types>`)
	return b.String()
}

func rootRels() string {
	return xmlDecl +
		`<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">` +
		`<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="xl/workbook.xml"/>` +
		`<Relationship Id="rIdGmd" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/customXml" Target="` + CarryPart + `"/>` +
		`</Relationships>`
}

func workbookXML(wb *model.Workbook) string {
	var b strings.Builder
	b.WriteString(xmlDecl)
	b.WriteString(`<workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"><sheets>`)
	for i, s := range wb.Sheets {
		state := ""
		switch s.Meta["hidden"] {
		case true:
			state = ` state="hidden"`
		case "very":
			state = ` state="veryHidden"`
		}
		fmt.Fprintf(&b, `<sheet name="%s" sheetId="%d"%s r:id="rId%d"/>`, xmlEscape(s.Name), i+1, state, i+1)
	}
	b.WriteString(`</sheets></workbook>`)
	return b.String()
}

func workbookRels(nSheets int) string {
	var b strings.Builder
	b.WriteString(xmlDecl)
	b.WriteString(`<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">`)
	for i := 0; i < nSheets; i++ {
		fmt.Fprintf(&b, `<Relationship Id="rId%d" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" Target="worksheets/sheet%d.xml"/>`, i+1, i+1)
	}
	fmt.Fprintf(&b, `<Relationship Id="rId%d" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles" Target="styles.xml"/>`, nSheets+1)
	b.WriteString(`</Relationships>`)
	return b.String()
}

func stylesXML() string {
	return xmlDecl +
		`<styleSheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main">` +
		`<fonts count="1"><font><sz val="11"/><name val="Calibri"/></font></fonts>` +
		`<fills count="1"><fill><patternFill patternType="none"/></fill></fills>` +
		`<borders count="1"><border/></borders>` +
		`<cellStyleXfs count="1"><xf numFmtId="0" fontId="0" fillId="0" borderId="0"/></cellStyleXfs>` +
		`<cellXfs count="2"><xf numFmtId="0" fontId="0" fillId="0" borderId="0" xfId="0"/>` +
		`<xf numFmtId="14" fontId="0" fillId="0" borderId="0" xfId="0" applyNumberFormat="1"/></cellXfs>` +
		`</styleSheet>`
}

func worksheetXML(s *model.Sheet, dateSystem int) (string, int) {
	// Group content-bearing cells by row.
	rows := map[int][]*model.Cell{}
	var rowNums []int
	count := 0
	for _, c := range s.Cells() {
		if _, ok := rows[c.Row]; !ok {
			rowNums = append(rowNums, c.Row)
		}
		rows[c.Row] = append(rows[c.Row], c)
	}
	sortInts(rowNums)

	var b strings.Builder
	b.WriteString(xmlDecl)
	b.WriteString(`<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><sheetData>`)
	for _, r := range rowNums {
		cells := rows[r]
		sortCellsByCol(cells)
		fmt.Fprintf(&b, `<row r="%d">`, r)
		for _, c := range cells {
			if writeCell(&b, c, dateSystem) {
				count++
			}
		}
		b.WriteString(`</row>`)
	}
	b.WriteString(`</sheetData>`)
	if len(s.Merges) > 0 {
		fmt.Fprintf(&b, `<mergeCells count="%d">`, len(s.Merges))
		for _, m := range s.Merges {
			ref := refs.NumToCol(m.C1) + strconv.Itoa(m.R1) + ":" + refs.NumToCol(m.C2) + strconv.Itoa(m.R2)
			fmt.Fprintf(&b, `<mergeCell ref="%s"/>`, ref)
		}
		b.WriteString(`</mergeCells>`)
	}
	b.WriteString(`</worksheet>`)
	return b.String(), count
}

// writeCell emits one cell; returns false when there is nothing representable.
func writeCell(b *strings.Builder, c *model.Cell, dateSystem int) bool {
	ref := refs.NumToCol(c.Col) + strconv.Itoa(c.Row)
	ct := c.Content
	switch {
	case ct.Rich != nil:
		var text strings.Builder
		for _, run := range ct.Rich {
			if s, ok := run["text"].(string); ok {
				text.WriteString(s)
			}
		}
		writeInlineStr(b, ref, text.String())
		return true
	case ct.Formula != nil:
		fmt.Fprintf(b, `<c r="%s"`, ref)
		if t := cachedType(ct.Cached); t != "" {
			fmt.Fprintf(b, ` t="%s"`, t)
		}
		b.WriteByte('>')
		fmt.Fprintf(b, `<f>%s</f>`, xmlEscape(*ct.Formula))
		if v := cachedValue(ct.Cached, dateSystem); v != "" {
			fmt.Fprintf(b, `<v>%s</v>`, v)
		}
		b.WriteString(`</c>`)
		return true
	case ct.Scalar != nil:
		return writeScalarCell(b, ref, ct.Scalar, dateSystem)
	default:
		return false
	}
}

func writeScalarCell(b *strings.Builder, ref string, sc *scalar.Scalar, dateSystem int) bool {
	switch sc.Kind {
	case scalar.Number:
		fmt.Fprintf(b, `<c r="%s"><v>%s</v></c>`, ref, numfmt.Format(sc.Num))
	case scalar.Boolean:
		v := "0"
		if sc.Bool {
			v = "1"
		}
		fmt.Fprintf(b, `<c r="%s" t="b"><v>%s</v></c>`, ref, v)
	case scalar.Error:
		fmt.Fprintf(b, `<c r="%s" t="e"><v>%s</v></c>`, ref, xmlEscape(sc.Str))
	case scalar.Date, scalar.Time:
		fmt.Fprintf(b, `<c r="%s" s="1"><v>%s</v></c>`, ref, numfmt.Format(isoToSerial(sc.Str, dateSystem)))
	case scalar.Text:
		writeInlineStr(b, ref, sc.Str)
	default:
		return false
	}
	return true
}

func writeInlineStr(b *strings.Builder, ref, text string) {
	fmt.Fprintf(b, `<c r="%s" t="inlineStr"><is><t xml:space="preserve">%s</t></is></c>`, ref, xmlEscape(text))
}

func cachedType(sc *scalar.Scalar) string {
	if sc == nil {
		return ""
	}
	switch sc.Kind {
	case scalar.Text:
		return "str"
	case scalar.Boolean:
		return "b"
	case scalar.Error:
		return "e"
	default:
		return ""
	}
}

func cachedValue(sc *scalar.Scalar, dateSystem int) string {
	if sc == nil {
		return ""
	}
	switch sc.Kind {
	case scalar.Number:
		return numfmt.Format(sc.Num)
	case scalar.Boolean:
		if sc.Bool {
			return "1"
		}
		return "0"
	case scalar.Error:
		return xmlEscape(sc.Str)
	case scalar.Date, scalar.Time:
		return numfmt.Format(isoToSerial(sc.Str, dateSystem))
	case scalar.Text:
		return xmlEscape(sc.Str)
	default:
		return ""
	}
}

func carryXML(b64 string) string {
	return xmlDecl +
		`<gridmd xmlns="urn:gridmd:carry" version="0.1"><source encoding="base64">` +
		b64 + `</source></gridmd>`
}

const xmlDecl = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>` + "\n"

func xmlEscape(s string) string {
	var b strings.Builder
	for _, r := range s {
		switch r {
		case '&':
			b.WriteString("&amp;")
		case '<':
			b.WriteString("&lt;")
		case '>':
			b.WriteString("&gt;")
		case '"':
			b.WriteString("&quot;")
		case '\'':
			b.WriteString("&apos;")
		default:
			b.WriteRune(r)
		}
	}
	return b.String()
}

func sortInts(a []int) {
	for i := 1; i < len(a); i++ {
		for j := i; j > 0 && a[j-1] > a[j]; j-- {
			a[j-1], a[j] = a[j], a[j-1]
		}
	}
}

func sortCellsByCol(a []*model.Cell) {
	for i := 1; i < len(a); i++ {
		for j := i; j > 0 && a[j-1].Col > a[j].Col; j-- {
			a[j-1], a[j] = a[j], a[j-1]
		}
	}
}

const dayMS = 86400000

// isoToSerial converts an ISO date/time (SPEC §6 rule 8) to an Excel serial per
// the date system (INTEROP §2), reproducing the 1900 phantom-leap-day offset.
func isoToSerial(iso string, dateSystem int) float64 {
	var datePart, timePart string
	if len(iso) >= 3 && iso[2] == ':' {
		timePart = iso
	} else if i := strings.IndexByte(iso, 'T'); i != -1 {
		datePart, timePart = iso[:i], iso[i+1:]
	} else {
		datePart = iso
	}

	frac := 0.0
	if timePart != "" {
		var hh, mm, ss int
		bits := strings.Split(timePart, ":")
		hh = atoi(bits[0])
		if len(bits) > 1 {
			mm = atoi(bits[1])
		}
		if len(bits) > 2 {
			ss = atoi(bits[2])
		}
		frac = float64(hh*3600+mm*60+ss) / 86400
	}
	if datePart == "" {
		return frac
	}

	ymd := strings.Split(datePart, "-")
	y, m, d := atoi(ymd[0]), atoi(ymd[1]), atoi(ymd[2])
	utc := time.Date(y, time.Month(m), d, 0, 0, 0, 0, time.UTC)
	var days float64
	if dateSystem == 1904 {
		epoch := time.Date(1904, 1, 1, 0, 0, 0, 0, time.UTC)
		days = float64(utc.Sub(epoch).Milliseconds()) / dayMS
	} else {
		epoch := time.Date(1899, 12, 30, 0, 0, 0, 0, time.UTC)
		days = float64(utc.Sub(epoch).Milliseconds()) / dayMS
		if days < 61 {
			days -= 1
		}
	}
	return days + frac
}

func atoi(s string) int {
	n, _ := strconv.Atoi(strings.TrimSpace(s))
	return n
}

File: rust/src/xlsx/write.rs
//! GridMD → XLSX. Emits a genuine worksheet core (cells, formulas + cached
//! values, merges, hidden state) natively, and carries the FULL original GridMD
//! source in a custom package part (`gridmd/source.gmd`) so nothing is ever
//! dropped (SPEC §11) and the round-trip is byte-exact. The importer restores
//! the model from that carry part (see `read.rs`); the native cells make the
//! file a real, openable spreadsheet.

use crate::dump::format_number;
use crate::model::{CellModel, Content, Model, SheetModel};
use crate::refs::num_to_col;
use crate::scalar::Scalar;
use crate::xlsx::zip::{zip_write, ZipEntry};
use crate::yaml::Yaml;

pub const CARRY_PART: &str = "gridmd/source.gmd";

/// ISO date/time → Excel serial per date-system (INTEROP §2). Port of
/// `units.js` `isoToSerial`, including the 1900 phantom-leap-day correction.
pub fn iso_to_serial(iso: &str, date_system: i64) -> f64 {
    let (date_part, time_part): (Option<&str>, Option<&str>) =
        if iso.len() >= 3 && &iso[2..3] == ":" {
            (None, Some(iso))
        } else {
            match iso.split_once('T') {
                Some((d, t)) => (Some(d), Some(t)),
                None => (Some(iso), None),
            }
        };
    let mut frac = 0.0;
    if let Some(t) = time_part {
        let mut it = t.split(':');
        let hh: f64 = it.next().and_then(|x| x.parse().ok()).unwrap_or(0.0);
        let mm: f64 = it.next().and_then(|x| x.parse().ok()).unwrap_or(0.0);
        let ss: f64 = it.next().and_then(|x| x.parse().ok()).unwrap_or(0.0);
        frac = (hh * 3600.0 + mm * 60.0 + ss) / 86400.0;
    }
    let date_part = match date_part {
        Some(d) => d,
        None => return frac,
    };
    let mut parts = date_part.split('-');
    let y: i64 = parts.next().and_then(|x| x.parse().ok()).unwrap_or(1900);
    let m: i64 = parts.next().and_then(|x| x.parse().ok()).unwrap_or(1);
    let d: i64 = parts.next().and_then(|x| x.parse().ok()).unwrap_or(1);
    let utc = days_from_civil(y, m, d);
    let mut days = if date_system == 1904 {
        (utc - days_from_civil(1904, 1, 1)) as f64
    } else {
        let dd = (utc - days_from_civil(1899, 12, 30)) as f64;
        if dd < 61.0 {
            dd - 1.0
        } else {
            dd
        }
    };
    days += frac;
    days
}

/// Days since the Unix epoch (proleptic Gregorian) — Howard Hinnant's algorithm.
fn days_from_civil(y: i64, m: i64, d: i64) -> i64 {
    let y = if m <= 2 { y - 1 } else { y };
    let era = if y >= 0 { y } else { y - 399 } / 400;
    let yoe = (y - era * 400) as i64;
    let doy = (153 * (if m > 2 { m - 3 } else { m + 9 }) + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146097 + doe - 719468
}

fn xml_escape(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for ch in s.chars() {
        match ch {
            '&' => out.push_str("&amp;"),
            '<' => out.push_str("&lt;"),
            '>' => out.push_str("&gt;"),
            '"' => out.push_str("&quot;"),
            '\'' => out.push_str("&apos;"),
            c => out.push(c),
        }
    }
    out
}

fn scalar_display(s: &Scalar) -> String {
    match s {
        Scalar::Number(n) => format_number(*n),
        Scalar::Boolean(b) => {
            if *b {
                "TRUE".to_string()
            } else {
                "FALSE".to_string()
            }
        }
        Scalar::Error(e) => e.clone(),
        Scalar::Date(d) | Scalar::Time(d) => d.clone(),
        Scalar::Text { value, .. } => value.clone(),
        Scalar::Blank => String::new(),
        Scalar::Formula { formula, .. } => formula.clone(),
    }
}

struct SharedStrings {
    list: Vec<String>,
    index: std::collections::HashMap<String, usize>,
}

impl SharedStrings {
    fn new() -> Self {
        SharedStrings {
            list: Vec::new(),
            index: std::collections::HashMap::new(),
        }
    }
    fn intern(&mut self, s: &str) -> usize {
        if let Some(&i) = self.index.get(s) {
            return i;
        }
        let i = self.list.len();
        self.list.push(s.to_string());
        self.index.insert(s.to_string(), i);
        i
    }
}

fn cell_xml(c: &CellModel, ss: &mut SharedStrings, date_system: i64, out: &mut String) {
    let content = match &c.content {
        Some(c) => c,
        None => return,
    };
    let refr = format!("{}{}", num_to_col(c.col), c.row);
    if let Some(rich) = &content.rich {
        let idx = ss.intern(rich);
        out.push_str(&format!("<c r=\"{refr}\" t=\"s\"><v>{idx}</v></c>"));
        return;
    }
    if let Some(formula) = &content.formula {
        formula_cell_xml(&refr, formula, content, out);
        return;
    }
    match &content.scalar {
        Some(Scalar::Number(n)) => {
            out.push_str(&format!("<c r=\"{refr}\"><v>{}</v></c>", format_number(*n)));
        }
        Some(Scalar::Boolean(b)) => {
            out.push_str(&format!(
                "<c r=\"{refr}\" t=\"b\"><v>{}</v></c>",
                if *b { 1 } else { 0 }
            ));
        }
        Some(Scalar::Error(e)) => {
            out.push_str(&format!(
                "<c r=\"{refr}\" t=\"e\"><v>{}</v></c>",
                xml_escape(e)
            ));
        }
        Some(Scalar::Date(d)) | Some(Scalar::Time(d)) => {
            let serial = iso_to_serial(d, date_system);
            out.push_str(&format!(
                "<c r=\"{refr}\" s=\"1\"><v>{}</v></c>",
                format_number(serial)
            ));
        }
        Some(Scalar::Text { value, .. }) => {
            let idx = ss.intern(value);
            out.push_str(&format!("<c r=\"{refr}\" t=\"s\"><v>{idx}</v></c>"));
        }
        _ => {
            // Blank / defensive: emit an empty inline cell.
            out.push_str(&format!("<c r=\"{refr}\"/>"));
        }
    }
}

fn formula_cell_xml(refr: &str, formula: &str, content: &Content, out: &mut String) {
    let f = xml_escape(formula);
    match &content.cached {
        None => out.push_str(&format!("<c r=\"{refr}\"><f>{f}</f></c>")),
        Some(Scalar::Number(n)) => out.push_str(&format!(
            "<c r=\"{refr}\"><f>{f}</f><v>{}</v></c>",
            format_number(*n)
        )),
        Some(other) => {
            let disp = xml_escape(&scalar_display(other));
            out.push_str(&format!(
                "<c r=\"{refr}\" t=\"str\"><f>{f}</f><v>{disp}</v></c>"
            ));
        }
    }
}

fn worksheet_xml(s: &SheetModel, ss: &mut SharedStrings, date_system: i64) -> String {
    let mut cells: Vec<&CellModel> = s.cells.iter().filter(|c| c.content.is_some()).collect();
    cells.sort_by(|a, b| a.row.cmp(&b.row).then(a.col.cmp(&b.col)));

    let mut body = String::new();
    body.push_str(
        "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<worksheet xmlns=\"http://schemas.openxmlformats.org/spreadsheetml/2006/main\">",
    );
    // dimension
    if let (Some(first), Some(last)) = (cells.first(), cells.last()) {
        let min_col = cells.iter().map(|c| c.col).min().unwrap_or(1);
        let max_col = cells.iter().map(|c| c.col).max().unwrap_or(1);
        let dim = format!(
            "{}{}:{}{}",
            num_to_col(min_col),
            first.row,
            num_to_col(max_col),
            last.row
        );
        body.push_str(&format!("<dimension ref=\"{dim}\"/>"));
    } else {
        body.push_str("<dimension ref=\"A1\"/>");
    }
    body.push_str("<sheetData>");
    let mut i = 0;
    while i < cells.len() {
        let row = cells[i].row;
        body.push_str(&format!("<row r=\"{row}\">"));
        while i < cells.len() && cells[i].row == row {
            cell_xml(cells[i], ss, date_system, &mut body);
            i += 1;
        }
        body.push_str("</row>");
    }
    body.push_str("</sheetData>");
    if !s.merges.is_empty() {
        body.push_str(&format!("<mergeCells count=\"{}\">", s.merges.len()));
        for m in &s.merges {
            let refr = format!(
                "{}{}:{}{}",
                num_to_col(m.c1),
                m.r1,
                num_to_col(m.c2),
                m.r2
            );
            body.push_str(&format!("<mergeCell ref=\"{refr}\"/>"));
        }
        body.push_str("</mergeCells>");
    }
    body.push_str("</worksheet>");
    body
}

fn shared_strings_xml(ss: &SharedStrings) -> String {
    let mut out = format!(
        "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<sst xmlns=\"http://schemas.openxmlformats.org/spreadsheetml/2006/main\" count=\"{0}\" uniqueCount=\"{0}\">",
        ss.list.len()
    );
    for s in &ss.list {
        out.push_str(&format!("<si><t xml:space=\"preserve\">{}</t></si>", xml_escape(s)));
    }
    out.push_str("</sst>");
    out
}

fn styles_xml() -> String {
    // One date cell format (numFmtId 14) at cellXfs index 1.
    "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<styleSheet xmlns=\"http://schemas.openxmlformats.org/spreadsheetml/2006/main\"><fonts count=\"1\"><font><sz val=\"11\"/><name val=\"Calibri\"/></font></fonts><fills count=\"1\"><fill><patternFill patternType=\"none\"/></fill></fills><borders count=\"1\"><border/></borders><cellStyleXfs count=\"1\"><xf numFmtId=\"0\" fontId=\"0\" fillId=\"0\" borderId=\"0\"/></cellStyleXfs><cellXfs count=\"2\"><xf numFmtId=\"0\" fontId=\"0\" fillId=\"0\" borderId=\"0\" xfId=\"0\"/><xf numFmtId=\"14\" fontId=\"0\" fillId=\"0\" borderId=\"0\" xfId=\"0\" applyNumberFormat=\"1\"/></cellXfs><cellStyles count=\"1\"><cellStyle name=\"Normal\" xfId=\"0\" builtinId=\"0\"/></cellStyles></styleSheet>".to_string()
}

fn workbook_xml(model: &Model) -> String {
    let mut out = String::from(
        "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<workbook xmlns=\"http://schemas.openxmlformats.org/spreadsheetml/2006/main\" xmlns:r=\"http://schemas.openxmlformats.org/officeDocument/2006/relationships\"><sheets>",
    );
    for (i, s) in model.sheets.iter().enumerate() {
        let state = match s.meta.get("hidden") {
            Some(Yaml::Bool(true)) => " state=\"hidden\"",
            Some(Yaml::Str(v)) if v == "very" => " state=\"veryHidden\"",
            _ => "",
        };
        out.push_str(&format!(
            "<sheet name=\"{}\" sheetId=\"{}\" r:id=\"rId{}\"{}/>",
            xml_escape(&s.name),
            i + 1,
            i + 1,
            state
        ));
    }
    out.push_str("</sheets></workbook>");
    out
}

fn workbook_rels_xml(model: &Model) -> String {
    let mut out = String::from(
        "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<Relationships xmlns=\"http://schemas.openxmlformats.org/package/2006/relationships\">",
    );
    for i in 0..model.sheets.len() {
        out.push_str(&format!(
            "<Relationship Id=\"rId{}\" Type=\"http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet\" Target=\"worksheets/sheet{}.xml\"/>",
            i + 1,
            i + 1
        ));
    }
    let n = model.sheets.len();
    out.push_str(&format!(
        "<Relationship Id=\"rId{}\" Type=\"http://schemas.openxmlformats.org/officeDocument/2006/relationships/sharedStrings\" Target=\"sharedStrings.xml\"/>",
        n + 1
    ));
    out.push_str(&format!(
        "<Relationship Id=\"rId{}\" Type=\"http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles\" Target=\"styles.xml\"/>",
        n + 2
    ));
    out.push_str("</Relationships>");
    out
}

fn content_types_xml(model: &Model) -> String {
    let mut out = String::from(
        "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<Types xmlns=\"http://schemas.openxmlformats.org/package/2006/content-types\"><Default Extension=\"rels\" ContentType=\"application/vnd.openxmlformats-package.relationships+xml\"/><Default Extension=\"xml\" ContentType=\"application/xml\"/><Default Extension=\"gmd\" ContentType=\"text/gridmd\"/><Override PartName=\"/xl/workbook.xml\" ContentType=\"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml\"/>",
    );
    for i in 0..model.sheets.len() {
        out.push_str(&format!(
            "<Override PartName=\"/xl/worksheets/sheet{}.xml\" ContentType=\"application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml\"/>",
            i + 1
        ));
    }
    out.push_str("<Override PartName=\"/xl/sharedStrings.xml\" ContentType=\"application/vnd.openxmlformats-officedocument.spreadsheetml.sharedStrings+xml\"/>");
    out.push_str("<Override PartName=\"/xl/styles.xml\" ContentType=\"application/vnd.openxmlformats-officedocument.spreadsheetml.styles+xml\"/>");
    out.push_str("</Types>");
    out
}

fn root_rels_xml() -> String {
    "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<Relationships xmlns=\"http://schemas.openxmlformats.org/package/2006/relationships\"><Relationship Id=\"rId1\" Type=\"http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument\" Target=\"xl/workbook.xml\"/></Relationships>".to_string()
}

/// The zip parts of the package, optionally including the GridMD carry part.
pub fn build_parts(model: &Model, source: &str, include_carry: bool) -> Vec<ZipEntry> {
    let date_system = if model.fm.get("date-system").and_then(|d| d.as_i64()) == Some(1904) {
        1904
    } else {
        1900
    };
    let mut ss = SharedStrings::new();
    let mut worksheets: Vec<(String, String)> = Vec::new();
    for (i, s) in model.sheets.iter().enumerate() {
        let xml = worksheet_xml(s, &mut ss, date_system);
        worksheets.push((format!("xl/worksheets/sheet{}.xml", i + 1), xml));
    }

    let mut entries = vec![
        ZipEntry {
            name: "[Content_Types].xml".to_string(),
            data: content_types_xml(model).into_bytes(),
        },
        ZipEntry {
            name: "_rels/.rels".to_string(),
            data: root_rels_xml().into_bytes(),
        },
        ZipEntry {
            name: "xl/workbook.xml".to_string(),
            data: workbook_xml(model).into_bytes(),
        },
        ZipEntry {
            name: "xl/_rels/workbook.xml.rels".to_string(),
            data: workbook_rels_xml(model).into_bytes(),
        },
        ZipEntry {
            name: "xl/styles.xml".to_string(),
            data: styles_xml().into_bytes(),
        },
    ];
    for (name, xml) in worksheets {
        entries.push(ZipEntry {
            name,
            data: xml.into_bytes(),
        });
    }
    entries.push(ZipEntry {
        name: "xl/sharedStrings.xml".to_string(),
        data: shared_strings_xml(&ss).into_bytes(),
    });
    if include_carry {
        entries.push(ZipEntry {
            name: CARRY_PART.to_string(),
            data: source.as_bytes().to_vec(),
        });
    }
    entries
}

/// A native-cell count per sheet, for the loud fidelity report.
pub fn native_cell_count(model: &Model) -> usize {
    model
        .sheets
        .iter()
        .map(|s| s.cells.iter().filter(|c| c.content.is_some()).count())
        .sum()
}

/// Export a model + its source to an `.xlsx` byte buffer.
pub fn write_xlsx(model: &Model, source: &str) -> Vec<u8> {
    zip_write(&build_parts(model, source, true))
}
